docs/Migration Guide.md

```diff
@@ -0,0 +1,152 @@
+  and `JS::CompartmentOptions` to `JS::RealmOptions`.
+- In most cases, rename `JSCompartment` to `JS::Realm`.
+  In a few APIs such as weak pointer callbacks where compartments and
+  not realms are used, rename it instead to `JS::Compartment`.
+- Change APIs such as `JS_GetFunctionPrototype()` which previously
+  operated explicitly on a global object, to the new APIs that operate
+  implicitly on the global object of the current realm (in this example,
+  `JS::GetRealmFunctionPrototype()`.)
+
+### Use UTF8-aware compilation and evaluation ###
+
+In ESR 68, the `JS::Evaluate()` and `JS::Compile()` family of APIs have
+been simplified and made UTF8-aware.
+Instead of C strings of `char*` and `char16_t*`, they now take an
+instance of `JS::SourceText<mozilla::Utf8Unit>` or
+`JS::SourceText<char16_t>`, respectively, which in addition to making
+the text encoding explicit, also carries information about the ownership
+of the source text memory.
+
+**Recommendations:**
+- Create an instance of `JS::SourceText` and put your code in it using
+  its `init()` method.
+  Note that the `init()` method can fail.
+- Pass the `JS::SourceText` instance to `JS::Compile()` or
+  `JS::Evaluate()` instead of a C string and its length.
+- Remove any usage of `JS::CompileOptions::setUTF8()`, which is now made
+  obsolete by the type of `JS::SourceText` passed in.
+- A few other APIs have been renamed for clarity, such as
+  `JS_BufferIsCompilableUnit()` to `JS_UTF8BufferIsCompilableUnit()`.
+- If using a source hook to manage in-memory sources,
+  `js::SourceHook::load()` now has two out parameters for either UTF8
+  source code or wide char source code.
+  Use only one of them.
+  (None of these examples currently use source hooks.)
+- Your code will need to include `<js/SourceText.h>` for
+  `JS::SourceText`.
+
+### String encoding ###
+
+`JS_EncodeString()` has been replaced with `JS_EncodeStringToASCII()`
+and `JS_EncodeStringToUTF8()`, so that the encoding of the returned
+string is always clear.
+
+**Recommendations:**
+- Depending on what you were using `JS_EncodeString()` for, replace it
+  with either the ASCII or UTF8 version.
+  For better performance in cases where you know the string only
+  contains ASCII characters, use the ASCII version, otherwise use the
+  UTF8 version.
+- These functions now return a `JS::UniqueChars` smart pointer, so there
+  is no longer any need to `JS_free()` the return value.
+- Where possible, change any of your internal APIs where the return
+  value of `JS_EncodeString()` was returned, to also return a smart
+  pointer.
+
+### Hash tables ###
+
+It is no longer necessary to initialize `JS::GCHashMap<>` with `init()`.
+
+**Recommendation:** Simply remove the call to `init()`.
+
+### Principals ###
+
+Functions having to do with stack frames, such as
+`JS::GetSavedFrameSource()` and others, now take a `JSPrincipals*`
+pointer as an argument.
+**FIXME**: What is that for?
+
+**Recommendations:**
+- **FIXME**
```

Comment:
If you are not doing sophisticated things with principals, you can find the current context's principal with `cx->realm()->principals();`